rbac/management/workspace/model.py

```diff
@@ -0,0 +1,40 @@
+#
+# Copyright 2024 Red Hat, Inc.
+#
+# This program is free software: you can redistribute it and/or modify
+# it under the terms of the GNU Affero General Public License as
+# published by the Free Software Foundation, either version 3 of the
+# License, or (at your option) any later version.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU Affero General Public License for more details.
+#
+# You should have received a copy of the GNU Affero General Public License
+# along with this program.  If not, see <https://www.gnu.org/licenses/>.
+#
+"""Model for workspace management."""
+from uuid import uuid4
+
+from django.db import models
+from django.utils import timezone
+from management.rbac_fields import AutoDateTimeField
+
+from api.models import TenantAwareModel
+
+
+class Workspace(TenantAwareModel):
+    """A workspace."""
+
+    name = models.CharField(max_length=100)
+    uuid = models.UUIDField(default=uuid4, editable=False, unique=True, null=False)
+    description = models.TextField(null=True)
+    created = models.DateTimeField(default=timezone.now)
+    modified = AutoDateTimeField(default=timezone.now)
```

Comment:
What fields do inventory groups have, for comparison?

Reply:
We should also reference their validation rules (e.g. if they allow names longer than 200 chars, we could have a migration problem).

Reply:
Another note: we may want to add an integer `version` field that gets incremented with every update.

I'm still noodling on this. But the idea would be to assist in synchronizing spicedb & RBAC db's accurately, without out-of-order updates. It would also help with just end user concurrency control (e.g. two users try to update a workspace at the same time).

It may not be necessary. It depends on how we want to synchronize data. I'll work on this some more and get back to you.

Reply:
@alechenninger 

Inventory Groups:
DB:
id - uuid
org_id - varchar(36) 
account - varchar(10)
name - varchar(255)
created_on - TIMESTAMP WITH TZ
modified_on - TIMESTAMP WITH TZ


API:
```
   {
      "id": "f34402bf-a002-49d4-b1d6-e89afd18e6fa",
      "org_id": "11789772",
      "account": "6089719",
      "name": "abc_237464d5-890c-4a4c-ae4a-dc8ecda04637",
      "host_count": 1, (calculate in their python app)
      "created": "2024-03-27T12:31:16.441718+00:00",
      "updated": "2024-04-10T09:12:24.313976+00:00"
    }
```

[HBI Group Model](https://github.com/RedHatInsights/insights-host-inventory/blob/0c8cb31702b9d8c3fd91bb139afdab91346b3b84/app/models.py#L506-L512)
[HBI Group Migration](https://github.com/RedHatInsights/insights-host-inventory/blob/master/migrations/versions/ada40da60e4e_add_groups_and_hosts_groups_tables.py#L22)
[HBI group API](https://console.redhat.com/docs/api/inventory/v1#operations-groups-api\.group\.get_group_list())
[HBI group validation for creation](https://github.com/RedHatInsights/insights-host-inventory/blob/master/app/models.py#L885)

Reply:
@alechenninger 

Based on that I will change length of name field to 255.

DB changes about sync - I would do this in follow up PR - once we will discuss this(I think we have meeting in calendar about  it)

OrgID/Account is stored in separated table in api_tenant as there is reference for it workspace#tenant_id (thanks to TenantAwareModel )


As I understand we need also migrate hosts to DB or just to Relation API ? Anyway If so we can also add it in followup - there M:N between inv. group and hosts - so it would need [joining table](https://github.com/RedHatInsights/insights-host-inventory/blob/master/migrations/versions/ada40da60e4e_add_groups_and_hosts_groups_tables.py#L36).

Reply:
> As I understand we need also migrate hosts to DB or just to Relation API ? Anyway If so we can also add it in followup - there M:N between inv. group and hosts - so it would need joining table.

Hosts: Just to Relations API. We may need to migrate more for Common Inventory but TBD.

Inventory Groups: To both RBAC DB and Relations API.

> There is uuid on the workspace model,
> 
> I believe we should also another column for inventory group uuid - like inventory_group_uuid.

I think we might rather want to migrate those UUIDs as is for compatibility. Is there any reason to avoid reusing those UUIDs? That way Workspace API can be forwarded to directly from HBI and not have to include two ways of referring to a Workspace.

**Additional note on validation**

I would like to consider adding a _minimum_ Workspace name length. When searching for a workspace, if you need to find a workspace that's only named "a" then this is somewhat of an operational problem because we have to pull every workspace that matches a very generic criteria. If you know every workspace is at least 3-4 letters or whatever, then you would never need to specify less for filters.

The problem is this might conflict with existing Inventory Groups, since looking through their code I do not see a minimum enforced. Is it possible to find out what the current minimum Inventory Group name length is in practice?

It's not a hard requirement I'd say but something we might thank ourselves for later.

Reply:
> Shared vs separated UUID
> My the only point is possible conflict due to same UUIDs for our UUIDs for non IG workspace and theirs.
> What if they change this uuid identification ?
> Basically we would rely on their UUIDs generation and their validations for it.

I would like to avoid this problem by basically doing a one time migration from their data, and then after which they no longer have any inventory group data of their only. We replace inventory groups with workspaces and therefore there isn't a separate inventory group ID any more. There is just workspaces which replaced inventory groups.